common/hashing.h

```diff
@@ -474,82 +495,141 @@ inline auto CarbonHashValue(const llvm::SmallString<Length>& value,
   return CarbonHashValue(llvm::StringRef(value.data(), value.size()), seed);
 }
 
-// C++ guarantees this is true for the unsigned variants, but we require it for
-// signed variants and pointers.
-static_assert(std::has_unique_object_representations_v<int8_t>);
-static_assert(std::has_unique_object_representations_v<int16_t>);
-static_assert(std::has_unique_object_representations_v<int32_t>);
-static_assert(std::has_unique_object_representations_v<int64_t>);
-static_assert(std::has_unique_object_representations_v<void*>);
-
-// C++ uses `std::nullptr_t` but unfortunately doesn't make it have a unique
-// object representation. To address that, we need a function that converts
-// `nullptr` back into a `void*` that will have a unique object representation.
-// And this needs to be done by-value as we need to build a temporary object to
-// return, which requires a separate overload rather than just using a type
-// function that could be used in parallel in the predicate below. Instead, we
-// build the predicate independently of the mapping overload, but together they
-// should produce the correct result.
+// Support types that are array like by building an `llvm::ArrayRef` out of
+// them. We can't do this by accepting any type convertible to an `ArrayRef`
+// because that type supports building a synthetic array out of any single
+// element. These need to go back through the dispatch implementation though to
+// pick up the full space of hash overloads for `T`.
```

Comment:
```suggestion
// element.
```
If we want to say this, I think we ought to do so on the `CarbonHashValue` overload for `ArrayRef`, but I think we can just drop it given that we don't call `DispatchImpl` either here or there.